RemoteImageBufferSet::prepareBufferForDisplay uses source rects larger than the layer bounds

kkinnunen-apple opened the pull request:
#### 34c2d1233dec1417e97868ce1c575e31b7f15fbc
<pre>
RemoteImageBufferSet::prepareBufferForDisplay uses source rects larger than the layer bounds
<a href="https://bugs.webkit.org/show_bug.cgi?id=266628">https://bugs.webkit.org/show_bug.cgi?id=266628</a>
<a href="https://rdar.apple.com/problem/119862411">rdar://problem/119862411</a>

Reviewed by Matt Woodrow.

The drawImageBuffer(ib, ..., sourceRect, ..) would receive
   FloatRect sourceRect = expandedIntRect({ { }, ib-&gt;logicalSize() })

This does not make sense, it is not well defined what it means to fetch
pixels outside of the source. This should be a caller bug.
For example, the logicalSize might be 95.5 with scale = 2. The
underlying native image would be 191 pixels.

The problem was caused by dirty rect calculations with Region, which
does not support floating point dimensions.

* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::prepareBufferForDisplay):

Canonical link: <a href="https://commits.webkit.org/272330@main">https://commits.webkit.org/272330@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ab494ec93b245cdd975cc92ba59493ad216cca4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31160 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32852 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33667 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28306 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12179 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7094 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27955 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31496 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8293 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27856 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7120 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7299 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35009 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28360 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28208 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33429 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7334 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5391 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31268 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9020 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7360 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8036 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow reviewed (approved):
Sad and unfortunate, but the same bug exists in RemoteLayerWithInProcessRenderingBackingStore::createContextAndPaintContents().

PR #21615 is going to merge the two (identical) code paths, so I can rebase that on top of these changes.

hortont424 commented on Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
```diff
@@ -181,17 +181,20 @@ void RemoteImageBufferSet::prepareBufferForDisplay(const WebCore::Region& dirtyR
+        IntRect enclosingCopyRect { m_previouslyPaintedRect ? *m_previouslyPaintedRect : enclosingIntRect(bufferBounds) };
+        if (!dirtyRegion.contains(enclosingCopyRect)) {
+            Region copyRegion(enclosingCopyRect);
+            copyRegion.subtract(dirtyRegion);
```
Don't you need to fix this same bug for in-process backing store? (Until Matt figures out how to share the code).

hortont424 replied:
LOL I see he said the same thing, ignore me.

kkinnunen-apple committed:
RemoteImageBufferSet::prepareBufferForDisplay uses source rects larger than the layer bounds
https://bugs.webkit.org/show_bug.cgi?id=266628
rdar://problem/119862411

Reviewed by Matt Woodrow.

The drawImageBuffer(ib, ..., sourceRect, ..) would receive
   FloatRect sourceRect = expandedIntRect({ { }, ib->logicalSize() })

This does not make sense, it is not well defined what it means to fetch
pixels outside of the source. This should be a caller bug.
For example, the logicalSize might be 95.5 with scale = 2. The
underlying native image would be 191 pixels.

The problem was caused by dirty rect calculations with Region, which
does not support floating point dimensions.

* Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp:
(WebKit::RemoteImageBufferSet::prepareBufferForDisplay):

Canonical link: https://commits.webkit.org/272330@main
